lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -1626,14 +1630,17 @@ def plot_surface(self, X, Y, Z, *args, norm=None, vmin=None,
                     ps = [
                         # +1 ensures we share edges between polygons
                         cbook._array_perimeter(a[rs:rs_next+1, cs:cs_next+1])
-                        for a in (X, Y, Z)
+                        for a in (X, Y, Z, mask)
                     ]
                     # ps = np.stack(ps, axis=-1)
                     ps = np.array(ps).T
                     polys.append(ps)
 
                     if fcolors is not None:
                         colset.append(fcolors[rs][cs])
+            # For the masking we have to loop because the polys may
+            # not be regularly shaped
+            polys = [p[:, :3] for p in polys if not np.any(p[:, 3])]
```

Comment:
I'm struggling to work out what this line of code is doing, I think because it's not obvious what `not np.any(p[:, 3])` is doing - could you expand the comment above a bit to explain this?